mito-ai/mito_ai/tests/providers/test_retry_logic.py

```diff
@@ -0,0 +1,386 @@
+from unittest.mock import AsyncMock, MagicMock, patch
+from mito_ai.completions.providers import OpenAIProvider
+from mito_ai.completions.models import MessageType, CompletionError
+from mito_ai.utils.mito_server_utils import ProviderCompletionException
+from mito_ai.tests.providers.utils import mock_openai_client, patch_server_limits
+from traitlets.config import Config
+
+FAKE_API_KEY = "sk-1234567890"
+
+@pytest.fixture
+def provider_config() -> Config:
+    """Create a proper Config object for the OpenAIProvider."""
+    config = Config()
+    config.OpenAIProvider = Config()
+    config.OpenAIClient = Config()
+    return config
+
+@pytest.fixture
+def mock_messages():
+    """Sample messages for testing."""
+    return [{"role": "user", "content": "Test message"}]
+
+@pytest.fixture
+def mock_sleep():
+    """Mock asyncio.sleep to avoid delays in tests."""
+    with patch("asyncio.sleep", new_callable=AsyncMock) as mock:
+        yield mock
+
+class TestRetryLogic:
+    """Test retry logic in OpenAIProvider.request_completions."""
+
+    @pytest.mark.parametrize("attempts_before_success,max_retries,expected_call_count", [
+        (0, 3, 1),  # Success on first try
+        (1, 3, 2),  # Success after 1 retry
+        (2, 3, 3),  # Success after 2 retries
+        (3, 3, 4),  # Success on final try
+    ])
+    @pytest.mark.asyncio
+    async def test_success_after_retries(
+        self, 
+        attempts_before_success: int, 
+        max_retries: int, 
+        expected_call_count: int,
+        provider_config: Config,
+        mock_messages,
+        mock_sleep,
+        monkeypatch: pytest.MonkeyPatch
+    ):
+        """Test that request_completions succeeds after a certain number of retries."""
+        monkeypatch.setenv("OPENAI_API_KEY", FAKE_API_KEY)
+        monkeypatch.setattr("mito_ai.constants.OPENAI_API_KEY", FAKE_API_KEY)
+        
+        # Create side effect: fail N times, then succeed
+        side_effects = [Exception("Test error")] * attempts_before_success + ["Success!"]
+        
+        with (
+            patch_server_limits(),
+            mock_openai_client() as mock_client
+        ):
+            mock_client.return_value.request_completions = AsyncMock(side_effect=side_effects)
+            
+            provider = OpenAIProvider(config=provider_config)
+            
+            # Test successful completion
+            result = await provider.request_completions(
+                message_type=MessageType.CHAT,
+                messages=mock_messages,
+                model="gpt-4o-mini",
+                max_retries=max_retries
+            )
+            
+            assert result == "Success!"
+            assert mock_client.return_value.request_completions.call_count == expected_call_count
+            assert provider.last_error is None  # Should be reset on success
+            
+            # Verify sleep was called for retries
+            assert mock_sleep.call_count == attempts_before_success
```

Comment:
Not important: The current tests only verify that asyncio.sleep is called the correct number of times, but they don't validate that the exponential backoff timing is implemented correctly.

Reply:
We can try something like this 

```suggestion
            expected_calls = [call(2**i) for i in range(attempts_before_success)]
            mock_sleep.assert_has_calls(expected_calls)

```